packages/astro/src/core/build/plugins/plugin-ssr.ts

```diff
@@ -1,62 +1,95 @@
-
+import type { AstroBuildPlugin } from '../plugin';
 import glob from 'fast-glob';
 import { fileURLToPath } from 'url';
 import { runHookBuildSsr } from '../../../integrations/index.js';
 import { isHybridOutput } from '../../../prerender/utils.js';
 import { BEFORE_HYDRATION_SCRIPT_ID, PAGE_SCRIPT_ID } from '../../../vite-plugin-scripts/index.js';
-import { pagesVirtualModuleId } from '../../app/index.js';
 import { joinPaths, prependForwardSlash } from '../../path.js';
 import { serializeRouteData } from '../../routing/index.js';
 import { addRollupInput } from '../add-rollup-input.js';
 import { getOutFile, getOutFolder } from '../common.js';
 import { cssOrder, mergeInlineCss, type BuildInternals } from '../internal.js';
-import type { AstroBuildPlugin } from '../plugin';
+import { MIDDLEWARE_MODULE_ID } from './plugin-middleware.js';
+import { extname } from 'node:path';
 import { RENDERERS_MODULE_ID } from './plugin-renderers.js';
+import { ASTRO_PAGE_EXTENSION_POST_PATTERN, ASTRO_PAGE_MODULE_ID } from './plugin-pages.js';
 
-export const virtualModuleId = '@astrojs-ssr-virtual-entry';
-const resolvedVirtualModuleId = '\0' + virtualModuleId;
+export const SSR_VIRTUAL_MODULE_ID = '@astrojs-ssr-virtual-entry';
+const RESOLVED_SSR_VIRTUAL_MODULE_ID = '\0' + SSR_VIRTUAL_MODULE_ID;
 const manifestReplace = '@@ASTRO_MANIFEST_REPLACE@@';
 const replaceExp = new RegExp(`['"](${manifestReplace})['"]`, 'g');
 
 function vitePluginSSR(
 	internals: BuildInternals,
 	adapter: AstroAdapter,
-	config: AstroConfig
+	options: StaticBuildOptions
 ): VitePlugin {
 	return {
 		name: '@astrojs/vite-plugin-astro-ssr',
 		enforce: 'post',
 		options(opts) {
-			return addRollupInput(opts, [virtualModuleId]);
+			return addRollupInput(opts, [SSR_VIRTUAL_MODULE_ID]);
 		},
 		resolveId(id) {
-			if (id === virtualModuleId) {
-				return resolvedVirtualModuleId;
+			if (id === SSR_VIRTUAL_MODULE_ID) {
+				return RESOLVED_SSR_VIRTUAL_MODULE_ID;
 			}
 		},
-		load(id) {
-			if (id === resolvedVirtualModuleId) {
-				let middleware = '';
+		async load(id) {
+			if (id === RESOLVED_SSR_VIRTUAL_MODULE_ID) {
+				const {
+					settings: { config },
+					allPages,
+				} = options;
+				const imports: string[] = [];
+				const contents: string[] = [];
+				const exports: string[] = [];
+				let middleware;
 				if (config.experimental?.middleware === true) {
-					middleware = 'middleware: _main.middleware';
+					imports.push(`import * as _middleware from "${MIDDLEWARE_MODULE_ID}"`);
+					middleware = 'middleware: _middleware';
+				}
+				let i = 0;
+				const pageMap: string[] = [];
+
+				for (const path of Object.keys(allPages)) {
+					const extension = extname(path);
+					const virtualModuleName = `${ASTRO_PAGE_MODULE_ID}${path.replace(
+						extension,
+						extension.replace('.', ASTRO_PAGE_EXTENSION_POST_PATTERN)
+					)}`;
+					let module = await this.resolve(virtualModuleName);
+					if (module) {
```

Comment:
It would be nice to document why we're constructing a module name this way, and how it'll ba handled later (in `entryFileNames`)

Reply:
I did, in the `README.md` I added

Reply:
With the function refactor I guess we can skip this then. It would be nice still since (I assume) the README is for higher level stuff, and inline code comments add the guard rails when reading.

Reply:
Done!